File: data/src/main/java/com/ortin/gesturetranslator/data/storage/SharedPrefSettingsStorage.java
package com.ortin.gesturetranslator.data.storage;

import android.content.Context;
import android.content.SharedPreferences;

import com.ortin.gesturetranslator.data.models.Settings;

public class SharedPrefSettingsStorage implements SettingsStorage {
    private static final String SHARED_PREFS_NAME = "settings";
    private static final String KEY_THEME = "theme";
    private static final String KEY_GPU = "gpu";
    private static final String KEY_PERCENT = "percent";
    private static final String KEY_SPEED_FRAME_DETECTION = "speed_frame_detection";

    private static final boolean DEFAULT_THEME = false;
    private static final boolean DEFAULT_GPU = false;
    private static final boolean DEFAULT_PERCENT = false;
    private static final int DEFAULT_SPEED_FRAME_DETECTION = 10;

    private final SharedPreferences sharedPreferences;

    public SharedPrefSettingsStorage(Context context) {
        sharedPreferences = context.getSharedPreferences(SHARED_PREFS_NAME, Context.MODE_PRIVATE);
    }

    @Override
    public boolean saveTheme(Settings settings) {
        sharedPreferences.edit().putBoolean(KEY_THEME, settings.isTheme()).apply();
        return true;
    }

    @Override
    public Settings getTheme() {
        Settings settings = new Settings();
        settings.setTheme(sharedPreferences.getBoolean(KEY_THEME, DEFAULT_THEME));
        return settings;
    }

    @Override
    public boolean saveGpu(Settings settings) {
        sharedPreferences.edit().putBoolean(KEY_GPU, settings.isGpu()).apply();
        return true;
    }

    @Override
    public Settings getGpu() {
        Settings settings = new Settings();
        settings.setGpu(sharedPreferences.getBoolean(KEY_GPU, DEFAULT_GPU));
        return settings;
    }

    @Override
    public boolean savePercent(Settings settings) {
        sharedPreferences.edit().putBoolean(KEY_PERCENT, settings.isPercent()).apply();
        return true;
    }

    @Override
    public Settings getPercent() {
        Settings settings = new Settings();
        settings.setPercent(sharedPreferences.getBoolean(KEY_PERCENT, DEFAULT_PERCENT));
        return settings;
    }

    @Override
    public boolean saveSpeedFrameDetection(Settings settings) {
        sharedPreferences.edit().putInt(KEY_SPEED_FRAME_DETECTION, settings.getSpeedFrameDetection()).apply();
        return true;
    }

    @Override
    public Settings getSpeedFrameDetection() {
        Settings settings = new Settings();
        settings.setSpeedFrameDetection(sharedPreferences.getInt(KEY_SPEED_FRAME_DETECTION, DEFAULT_SPEED_FRAME_DETECTION));
        return settings;
    }
}

File: data/src/main/kotlin/com/ortin/gesturetranslator/data/storage/SharedPrefSettingsStorage.kt
package com.ortin.gesturetranslator.data.storage

import android.content.Context
import android.content.SharedPreferences
import com.ortin.gesturetranslator.data.models.Settings
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject

class SharedPrefSettingsStorage @Inject constructor(@ApplicationContext context: Context) :
    SettingsStorage {
    private val sharedPreferences: SharedPreferences =
        context.getSharedPreferences(SHARED_PREFS_NAME, Context.MODE_PRIVATE)

    override fun saveSettings(settings: Settings): Boolean {
        sharedPreferences.edit().putBoolean(KEY_THEME, settings.theme).apply()
        sharedPreferences.edit().putBoolean(KEY_GPU, settings.gpu).apply()
        sharedPreferences.edit().putBoolean(KEY_PERCENT, settings.percent).apply()
        sharedPreferences.edit().putInt(KEY_SPEED_FRAME_DETECTION, settings.speedFrameDetection)
            .apply()
        return true
    }

    override fun getSettings(): Settings =
        Settings(
            theme = sharedPreferences.getBoolean(KEY_THEME, DEFAULT_THEME),
            gpu = sharedPreferences.getBoolean(KEY_GPU, DEFAULT_GPU),
            percent = sharedPreferences.getBoolean(KEY_PERCENT, DEFAULT_PERCENT),
            speedFrameDetection = sharedPreferences.getInt(
                KEY_SPEED_FRAME_DETECTION,
                DEFAULT_SPEED_FRAME_DETECTION
            )
        )

    companion object {
        private const val SHARED_PREFS_NAME: String = "settings"
        private const val KEY_THEME: String = "theme"
        private const val KEY_GPU: String = "gpu"
        private const val KEY_PERCENT: String = "percent"
        private const val KEY_SPEED_FRAME_DETECTION: String = "speed_frame_detection"
        private const val DEFAULT_THEME: Boolean = false
        private const val DEFAULT_GPU: Boolean = false
        private const val DEFAULT_PERCENT: Boolean = false
        private const val DEFAULT_SPEED_FRAME_DETECTION: Int = 10
    }
}
